### PR TITLE
Refactor payment methods adapter

### PR DESCRIPTION
**Pull request by caiofaustino**
## Summary
- Create view data models and refactor payment methods list generation on the DropInViewModel.
- Refactor PaymentMethodAdapter to separate regular and stored PaymentMethods.
- Remove expanded logic from PaymentMethodAdapter as it will be replaced with it's own fragment.

This if part of a bigger refactor, one click functionality still needs to be refactored.

### CODE REVIEW

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt [hunk 1]
```diff
@@ -10,84 +10,184 @@ package com.adyen.checkout.dropin.ui
...
+            }
+        }
+    }
+
+    private fun makeStoredModel(storedPaymentMethod: StoredPaymentMethod): StoredPaymentMethodModel {
+        return when (storedPaymentMethod.type) {
+            PaymentMethodTypes.SCHEME -> {
+                StoredPaymentMethodModel(
+                    storedPaymentMethod.id ?: "",
+                    "•••• ${storedPaymentMethod.lastFour}",
```
Comment (jreij) [inline drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt:102]: Should we do the formatting on the UI here instead? I can see we have `card_number_4digit` in the string resources and `CardView.setStoredCardInterface` already uses it

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt [hunk 2]
```diff
@@ -10,84 +10,184 @@ package com.adyen.checkout.dropin.ui
...
+                    "${storedPaymentMethod.expiryMonth}/${DateUtils.removeFirstTwoDigitFromYear(storedPaymentMethod.expiryYear)}",
```
Comment (jreij) [inline drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt:103]: Do we have a formatting util for this one? If not maybe we should extract one, or ideally let the UI do the formatting also. What do you think?

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt [hunk 3]
```diff
@@ -10,84 +10,184 @@ package com.adyen.checkout.dropin.ui
...
+            when {
+                type == null -> {
+                    throw CheckoutException("PaymentMethod type is null")
+                }
+                PaymentMethodTypes.SUPPORTED_PAYMENT_METHODS.contains(type) -> {
+                    Logger.v(TAG, "Supported payment method: $type")
+                    // We assume payment method is available and remove it later when the callback comes
+                    // this is the overwhelming majority of cases, and we keep the list ordered this way.
+                    paymentMethodsList.add(
+                        PaymentMethodModel(type, paymentMethod.name ?: nullNameFallback)
```
Comment (jreij) [inline drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt:139]: What about localization here? The UI will show "Unknown" for all languages. Or is this an edge case that shouldn't practically happen?

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt [hunk 4]
```diff
@@ -10,84 +10,184 @@ package com.adyen.checkout.dropin.ui
...
-                } else {
-                    Logger.e(TAG, "PaymentMethod not yet supported - $type")
+    override fun onAvailabilityResult(isAvailable: Boolean, paymentMethod: PaymentMethod, config: Configuration?) {
+        Logger.d(TAG, "onAvailabilityResult - ${paymentMethod.type}: $isAvailable")
+
+        availabilitySum++
+
+        if (!isAvailable) {
+            Logger.e(TAG, "${paymentMethod.type} NOT AVAILABLE")
+            for ((index, paymentMethodModel) in paymentMethodsList.withIndex()) {
```
Comment (jreij) [inline drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt:167]: We can do something like this: `paymentMethodsList.removeAll {  it.type == paymentMethod.type }` ?

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodAdapter.kt [hunk 5]
```diff
@@ -8,191 +8,150 @@
...
         }
     }
 
     override fun getItemViewType(position: Int): Int {
-        return when (getPaymentMethod(position)) {
-            HEADER_PLACEHOLDER_ITEM_P, HEADER_PLACEHOLDER_ITEM_SP -> PAYMENT_METHODS_HEADER
-            FOOTER_PLACEHOLDER_ITEM -> PAYMENT_METHODS_FOOTER
-            else -> PAYMENT_METHOD_ITEMS
+        // Top is always a header
+        return if (position == 0) {
```
Comment (jreij) [inline drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodAdapter.kt:45]: We can replace this with one `when` block?

```
    override fun getItemViewType(position: Int): Int {
        // Top is always a header
        return when {
            (position == 0) -> PAYMENT_METHODS_HEADER
            noStored() -> PAYMENT_METHOD // only regular payment methods left
            position <= paymentMethodsListModel.storedPaymentMethods.size -> STORED_PAYMENT_METHOD // 0 header - stored payments
            position == (paymentMethodsListModel.storedPaymentMethods.size + 1) -> PAYMENT_METHODS_HEADER //2nd header - regular payments
            else -> PAYMENT_METHOD
        }
    }
```

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodAdapter.kt [hunk 6]
```diff
@@ -8,191 +8,150 @@
...
+                    // 0 header - stored payments - 2nd header - regular payments
+                    position <= paymentMethodsListModel.storedPaymentMethods.size -> STORED_PAYMENT_METHOD
+                    position == (paymentMethodsListModel.storedPaymentMethods.size + 1) -> PAYMENT_METHODS_HEADER
+                    else -> PAYMENT_METHOD
+                }
+            }
         }
     }
 
     override fun onBindViewHolder(holder: BaseViewHolder, position: Int) {
```
Comment (jreij) [inline drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodAdapter.kt:62]: Can we extract 3 methods from this? Something like `bindPaymentMethodsHeaderVH` - `bindStoredPaymentMethodVH` - `bindPaymentMethodVH` ?

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt [hunk 7]
```diff
@@ -10,84 +10,184 @@ package com.adyen.checkout.dropin.ui
...
+    private var availabilitySkipSum = 0
+    private var availabilityChecksum = 0
+
+    private val storedPaymentMethodsList = mutableListOf<StoredPaymentMethodModel>()
+    private val paymentMethodsList = mutableListOf<PaymentMethodModel>()
 
-        if (isAvailable) {
-            addPaymentMethod(paymentMethod)
+    fun getStoredPaymentMethod(id: String): StoredPaymentMethod {
+        for (storedPaymentMethod in paymentMethodsApiResponse.storedPaymentMethods.orEmpty()) {
```
[the remark this answers is not in the document]
Comment (caiofaustino) [reply drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt:60]: Awesome, agreed!

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt (hunk 3 above)
Comment (caiofaustino) [reply drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt:139]: Good point, this shouldn't really happen, I was also considering to just leave "" for the fallback. Maybe that's better and the View can have the localized "Unknown".

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt (hunk 1 above)
Comment (caiofaustino) [reply drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt:102]: Yeah I was thinking the same thing, but I thought of having the `StoredPaymentMethodModel` generic for other payments that are not Card. Let's discuss offline.

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt (hunk 4 above)
Comment (caiofaustino) [reply drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt:167]: <3

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodAdapter.kt (hunk 5 above)
Comment (caiofaustino) [reply drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodAdapter.kt:45]: 👍

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodAdapter.kt (hunk 6 above)
Comment (caiofaustino) [reply drop-in/src/main/java/com/adyen/checkout/dropin/ui/paymentmethods/PaymentMethodAdapter.kt:62]: yeah I was considering that too, thanks

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt (hunk 2 above)
Comment (caiofaustino) [reply drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt:103]: Yep, let's do it.

File: drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt (hunk 1 above)
Comment (caiofaustino) [reply drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt:102]: As we discussed it's probably better to have typed `StoredPaymentMethodModel` sub classes and to parse the things in the View. I'll refactor this bit :)

**Review by jreij — APPROVED** [review]
👍

**Pull request merged**